db/inlineskiplist.h

```diff
@@ -549,43 +665,103 @@ void InlineSkipList<Comparator>::Insert(const char* key) {
       prev_[i] = prev_[0];
     }
   } else {
-    // TODO(opt): we could use a NoBarrier predecessor search as an
-    // optimization for architectures where memory_order_acquire needs
-    // a synchronization instruction.  Doesn't matter on x86
     FindLessThan(key, prev_);
   }
 
   // Our data structure does not allow duplicate insertion
   assert(prev_[0]->Next(0) == nullptr || !Equal(key, prev_[0]->Next(0)->Key()));
 
-  // Find the Node that we placed before the key in AllocateKey
-  Node* x = reinterpret_cast<Node*>(const_cast<char*>(key)) - 1;
-  int height = x->UnstashHeight();
-  assert(height >= 1 && height <= kMaxHeight_);
+  for (int i = 0; i < height; i++) {
+    x->InsertAfter(prev_[i], i);
+  }
+  prev_[0] = x;
+  prev_height_.store(height, std::memory_order_relaxed);
+}
 
-  if (height > GetMaxHeight()) {
-    for (int i = GetMaxHeight(); i < height; i++) {
-      prev_[i] = head_;
+template <class Comparator>
+void InlineSkipList<Comparator>::FindWithHint(
+    const char* key, InsertHint* hint, int start_level, int stop_level) const {
+  assert(start_level >= stop_level);
+  int level = start_level;
+  while (level < GetMaxHeight() &&
+      !KeyIsNextToNode(key, hint->prev[level], level)) {
+    level++;
+  }
+  if (level == start_level) {
+    while (level > stop_level &&
+        KeyIsNextToNode(key, hint->prev[level - 1], level - 1)) {
+      level--;
     }
-
-    // It is ok to mutate max_height_ without any synchronization
-    // with concurrent readers.  A concurrent reader that observes
-    // the new value of max_height_ will see either the old value of
-    // new level pointers from head_ (nullptr), or a new value set in
-    // the loop below.  In the former case the reader will
-    // immediately drop to the next level since nullptr sorts after all
-    // keys.  In the latter case the reader will use the new node.
-    max_height_.store(height, std::memory_order_relaxed);
   }
+  if (level > stop_level) {
+    FindLessThan(key, hint->prev, hint->prev[level], level, stop_level);
+  }
+}
```

Comment:
This function confused me. I see the function is used in two places and serve very different cases. Can we have two functions instead?